File: onibus/src/main/java/com/proj_db/onibus/service/OrdemServicoService.java
package com.proj_db.onibus.service;

import java.util.List;
import java.util.Optional;

import com.proj_db.onibus.dto.OrdemServicoCreateDTO;
import com.proj_db.onibus.dto.OrdemServicoSearchDTO;
import com.proj_db.onibus.dto.OrdemServicoUpdateDTO;
import com.proj_db.onibus.model.OrdemServico;

// <<< Imports para os DTOs externos que estão na pasta /dto
public interface OrdemServicoService {
    
    // --- CRUD e Ciclo de Vida ---
    OrdemServico create(OrdemServicoCreateDTO dto);
    OrdemServico updateInfo(Long osId, OrdemServicoUpdateDTO dto);
    void delete(Long osId); // Apenas se CANCELADA

    Optional<OrdemServico> findById(Long id);
    List<OrdemServico> findAll();
    List<OrdemServico> search(OrdemServicoSearchDTO criteria);

    // --- Ações Principais do Fluxo de Trabalho ---
    OrdemServico startExecution(Long osId);
    OrdemServico finishExecution(Long osId);
    OrdemServico cancel(Long osId);

    // --- Gerenciamento de Itens (através da OS) ---
    OrdemServico addItem(Long osId, Long produtoId, Double quantidade, String descricao);
    OrdemServico removeItem(Long osId, Long itemId);
    OrdemServico updateItemQuantity(Long osId, Long itemId, Double novaQuantidade);
    
}

File: frontend-onibus/src/services/ordemServicoService.js
import api from './api';

export const ordemServicoService = {
  // --- CRUD e Ciclo de Vida da OS ---
  create: (osCreateDTO) => api.post('/ordens-servico', osCreateDTO),
  updateInfo: (id, osUpdateDTO) => api.put(`/ordens-servico/${id}`, osUpdateDTO),
  delete: (id) => api.delete(`/ordens-servico/${id}`),
  
  // --- Consultas ---
  getAll: () => api.get('/ordens-servico'),
  getById: (id) => api.get(`/ordens-servico/${id}`),
  search: (terms) => api.get('/ordens-servico/search', { params: terms }),

  // --- Ações de Fluxo de Trabalho ---
  startExecution: (id) => api.patch(`/ordens-servico/${id}/start`),
  finishExecution: (id) => api.patch(`/ordens-servico/${id}/finish`),
  cancel: (id) => api.patch(`/ordens-servico/${id}/cancel`),

  // --- Gerenciamento de Itens (através da OS) ---
  addItem: (osId, itemDTO) => api.post(`/ordens-servico/${osId}/itens`, itemDTO),
  removeItem: (osId, itemId) => api.delete(`/ordens-servico/${osId}/itens/${itemId}`),
  updateItemQuantity: (osId, itemId, itemDTO) => api.put(`/ordens-servico/${osId}/itens/${itemId}`, itemDTO),
};

